k8s/contiv-vpp.yaml

```diff
@@ -254,22 +254,34 @@ spec:
         image: contivvpp/cni:latest
         imagePullPolicy: IfNotPresent
         env:
-          - name: SLEEP
-            value: "false"
+        - name: SLEEP
+          value: "false"
         volumeMounts:
-          - mountPath: /opt/cni/bin
-            name: cni-bin-dir
-          - mountPath: /etc/cni/net.d
-            name: cni-net-dir
-          - mountPath: /var/run/contiv
-            name: contiv-run
-      # This init container waits until etcd is started
-      - name: wait-foretcd
+        - mountPath: /opt/cni/bin
+          name: cni-bin-dir
+        - mountPath: /etc/cni/net.d
+          name: cni-net-dir
+        - mountPath: /var/run/contiv
+          name: contiv-run
+      # This init container waits until crd config has been completed
+      - name: wait-forcrdconfig
         env:
-          - name: ETCDPORT
-            value: "32379"
-        image: busybox:latest
-        command: ['sh', '-c', 'until nc -w 2 127.0.0.1:$ETCDPORT; do echo waiting for etcd; sleep 2; done;']
+        - name: ETCDPORT
+          value: "32379"
+        - name: ETCDCTL_API
+          value: "3"
+        image: quay.io/coreos/etcd:v3.1.10
+        command:
+        - /bin/sh
+        args:
+        - -c
+        - |
+          set -eu
+          until /usr/local/bin/etcdctl --endpoints=127.0.0.1:$ETCDPORT get "/vnf-agent/contiv-ksr/status/crd" --prefix=true | grep -m 1 "1";
```

Comment:
I will need to talk about this startup dependency with Lukas and Rasto once they are back from PTO. At this point I'm not sure if it is really needed + we want to be able to start node independently on the accessibility of ETCD eventually.

Reply:
I don't mind adding a mechanism in the yaml to check if contiv-crd agent is up and running, but we will need to find a way to resolve this as I think having CRD and KSR depending on ETCD and then other components depend on CRD/KSR readiness will probably cause problems in the future.

Reply:
Currently, the vswitch agent needs to wait for ETCD anyway, but as Milan mentioned, we want to make it independent on ETCD later on. For now, we could even go with a wait for CRD data in ETCD, but we still need to solve the startup dependency anyway: since the node config can come to CRD after CRD is started, how do we know if we should wait for it or not? I think we need some kind of a global switch - either CRD or YAML + auto-calculated defaults can be used for node config. We cannot support both at the same time. In case of CRD, the node config would be mandatory even if it is empty. Then the vswitch should always wait until its node config is available in CRD.

Reply:
If you have config data locally cached in Bolt, you can come up and resync with CRD later (i.e. if CRD config for you is different from the cached config, you have no choice but to restart with the new config). If there is no locally cached config, you have to wait for CRD. 

Now, if CRD is initially empty and we do want to keep supporting the initial config though the YAML file, we can have an init container that gets config from the YAML file and shoves it into CRD. We just have to decide what takes precedence in case of conflict - YAML or CRD.